### PR TITLE
im envy

### PR DESCRIPTION
**Pull request by envyaims**
_If any of the below don't apply to this Pull Request, mark the checkbox as done._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions), which include the following: 
  - I understand that if it is clear that I have not attempted to follow these guidelines (ex. if I have not used tabs to indent), my PR will be closed.
  - If changes are requested, I will re-request the review after making the changes.

### CODE REVIEW

File: solutions/advanced/cf-891C [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+		while (szsnap.size() > until) {
+			szsnap.back().first = szsnap.back().second;
+			szsnap.pop_back();
+			psnap.back().first = psnap.back().second;
+			psnap.pop_back();
+		}
+	}
+};
+
+struct Edge{
```
Comment (SansPapyrus683) [inline solutions/advanced/cf-891C:84]:
```suggestion
struct Edge {
```same for other things

File: solutions/advanced/cf-891C [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+	int u, v, w;
+};
+
+int main(){
+	cin.tie(0) -> sync_with_stdio(0);
```
Comment (SansPapyrus683) [inline solutions/advanced/cf-891C:89]: you know what to do

File: solutions/advanced/cf-891C [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+	vector<Edge> edges;
+	for (int i = 0; i < m; i++) {
+		int u, v, w;
+		cin >> u >> v >> w;
+		--u; --v;
+		edges.push_back({u, v, w});
+		edge_by_weight[w].push_back(i);
+	}
+	
+	int q; cin >> q;
```
Comment (SansPapyrus683) [inline solutions/advanced/cf-891C:101]: separate lines

File: solutions/advanced/cf-891C [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+	vector<bool> ans(q, true);
+	for (int i = 0; i < q; i++) {
+		int k; cin >> k;
+		for (int j = 0; j < k; j++) {
+			int x; cin >> x;
+			--x;
```
Comment (SansPapyrus683) [inline solutions/advanced/cf-891C:107]:
```suggestion
			x--;
```

File: solutions/advanced/cf-891C [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+			query_by_weight[edges[x].w][i].push_back(x);
+		}
+	}
+	
+	DSU dsu(n);
+	for (int weight = 1; weight < MAXN; weight++) {
+		for (auto& query: query_by_weight[weight]) {
```
Comment (SansPapyrus683) [inline solutions/advanced/cf-891C:114]: please don't

File: solutions/advanced/cf-891C [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+
+<LanguageSection>
+
+<CPPSection>
+
+```cpp 
+#include <bits/stdc++.h>
+using namespace std;
+
+const int MAXN = 5e5 + 5;
```
Comment (SansPapyrus683) [inline solutions/advanced/cf-891C:30]: does this work w/o +5?

File: solutions/advanced/cf-891C [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+source: CF
+title: Envy
+author: Chongtian Ma
+---
+
+[Official Analysis](https://codeforces.com/blog/entry/55841)
+
+## Explanation
+
+This problem requires some MST intuition. In Kruskal's algorithm, we are to sort the edges by non-decreasing weight, and try uniting vertices if they aren't already in the same component. Here, we can try to check every query using this algorithm, by processing each edge of each query offline.
```
Comment (SansPapyrus683) [inline solutions/advanced/cf-891C:12]: what does the first sentence contribute

File: solutions/advanced/cf-891C [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+
+As we consider edges from non-decreasing weight, we can keep track of a partially-built MST. An edge can only be in a MST if we processed all edges that have weights less than this edge, and we are able to unite the endpoints of the edge. This is because edges that have the same weight are arbitrary, so if we prioritize considering this edge then another MST with the same total weight is constructed.
+
+Thus, for each query, we need all edges to satisfy this condition. We can process all edges in non-decreasing weight, and for each query, try to unite the endpoint of all the edges that have this weight. If any edge does not satisfy, the whole query fails.
+
+## Explanation
+
+**Rough Time Complexity:** $\mathcal{O}(q \log n)$
```
Comment (SansPapyrus683) [inline solutions/advanced/cf-891C:20]: wdym "rough"

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: solutions/advanced/cf-891C (hunk 2 above)
Comment (envyaims) [reply solutions/advanced/cf-891C:89]: it alreayd runs on 1k out of 2k ms  you expect me to remove this and still runs in time

File: solutions/advanced/cf-891C (hunk 4 above)
Comment (envyaims) [reply solutions/advanced/cf-891C:107]: why do u not like

File: solutions/advanced/cf-891C (hunk 5 above)
Comment (envyaims) [reply solutions/advanced/cf-891C:114]: its too messy otherwise
plus i explained what everything is using variable names below already

File: solutions/advanced/cf-891C (hunk 6 above)
Comment (envyaims) [reply solutions/advanced/cf-891C:30]: want me to do +1?

File: solutions/advanced/cf-891C (hunk 7 above)
Comment (envyaims) [reply solutions/advanced/cf-891C:12]: that intuition is required to understand the following explanation

File: solutions/advanced/cf-891C (hunk 8 above)
Comment (envyaims) [reply solutions/advanced/cf-891C:20]: inverse ackerman and other things like maps which i didnt account for

Comment (dustin-miao) [reply solutions/advanced/cf-891C:20]: Assume DSU runs in O(1) and give the exact time complexity

File: solutions/advanced/cf-891C (hunk 4 above)
Comment (dustin-miao) [reply solutions/advanced/cf-891C:109]: If its not in the style guide there's no need to change it.

File: solutions/advanced/cf-891C (hunk 5 above)
Comment (dustin-miao) [reply solutions/advanced/cf-891C:116]: I don't see whats wrong with this

File: solutions/advanced/cf-891C [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+id: cf-891C
+source: CF
+title: Envy
+author: Chongtian Ma
+---
+
+[Official Analysis](https://codeforces.com/blog/entry/55841)
+
+## Explanation
+
```
Comment (dustin-miao) [inline solutions/advanced/cf-891C:11]: If an official editorial/solution exists you should provide a link

Comment (envyaims) [reply solutions/advanced/cf-891C:11]: ?? its literally there

File: solutions/advanced/cf-891C (hunk 2 above)
Comment (SansPapyrus683) [reply solutions/advanced/cf-891C:89]: so you're telling me io would account for an entire second of savings

File: solutions/advanced/cf-891C (hunk 4 above)
Comment (SansPapyrus683) [reply solutions/advanced/cf-891C:109]: fine, i'll click it myself

File: solutions/advanced/cf-891C (hunk 7 above)
Comment (SansPapyrus683) [reply solutions/advanced/cf-891C:12]: this problem is in advanced though
it's like saying "you need prefix sum intuition to solve this" in a gold solution

File: solutions/advanced/cf-891C (hunk 6 above)
Comment (SansPapyrus683) [reply solutions/advanced/cf-891C:30]: no, i want you to do +0

File: solutions/advanced/cf-891C (hunk 2 above)
Comment (envyaims) [reply solutions/advanced/cf-891C:89]: yes, since there are so many lines

File: solutions/advanced/cf-891C (hunk 6 above)
Comment (envyaims) [reply solutions/advanced/cf-891C:30]: i cant, unless you want me to put --s literally everywhere

File: solutions/advanced/cf-891C (hunk 7 above)
Comment (envyaims) [reply solutions/advanced/cf-891C]: should i put this as a hint instead?

Comment (SansPapyrus683) [reply solutions/advanced/cf-891C]: idt that would even qualify as a hint tbh

File: solutions/advanced/cf-891C (hunk 6 above)
Comment (SansPapyrus683) [reply solutions/advanced/cf-891C:30]: i don't even see a variable named s in your code

Comment (envyaims) [reply solutions/advanced/cf-891C:30]: i meant multiple "--"

Comment (SansPapyrus683) [reply solutions/advanced/cf-891C:30]: then do +1 and comment where needed "index 0 won't be used"

**Review by SansPapyrus683 — APPROVED** [review]
no one's gonna see this anyways

**Review by ChuangSheep — APPROVED** [review]
<del>if Sans approved it</del> it's all right

File: solutions/advanced/cf-891C [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+We can solve this using some MST intuition. In Kruskal's algorithm, we are to sort the edges by non-decreasing weight, and try uniting vertices if they aren't already in the same component. Here, we can try to check every query using this algorithm, by processing each edge of each query offline.
+
+As we consider edges from non-decreasing weight, we can keep track of a partially-built MST. An edge can only be in a MST if we processed all edges that have weights less than this edge, and we are able to unite the endpoints of the edge. This is because edges that have the same weight are arbitrary, so if we prioritize considering this edge then another MST with the same total weight is constructed.
+
+Thus, for each query, we need all edges to satisfy this condition. We can process all edges in non-decreasing weight, and for each query, try to unite the endpoint of all the edges that have this weight. If any edge does not satisfy, the whole query fails.
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(q \log n)$
```
Comment (dustin-miao) [inline solutions/advanced/cf-891C:20]: Update the complexity to the correct value

Comment (envyaims) [reply solutions/advanced/cf-891C:20]: is it not correct?

File: solutions/advanced/cf-891C [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+author: Chongtian Ma
+---
+
+[Official Analysis](https://codeforces.com/blog/entry/55841)
+
+## Explanation
+
+We can solve this using some MST intuition. In Kruskal's algorithm, we are to sort the edges by non-decreasing weight, and try uniting vertices if they aren't already in the same component. Here, we can try to check every query using this algorithm, by processing each edge of each query offline.
+
+As we consider edges from non-decreasing weight, we can keep track of a partially-built MST. An edge can only be in a MST if we processed all edges that have weights less than this edge, and we are able to unite the endpoints of the edge. This is because edges that have the same weight are arbitrary, so if we prioritize considering this edge then another MST with the same total weight is constructed.
```
[the remark this answers is not in the document]
Comment (envyaims) [reply solutions/advanced/cf-891C:14]: "higher chance" "current weight class"
wdym

File: solutions/advanced/cf-891C (hunk 10 above)
Comment (dustin-miao) [reply solutions/advanced/cf-891C:20]: There has to be a factor of $m\log m$ in there somewhere for the sorting right?

File: solutions/advanced/cf-891C (hunk 11 above)
[the remark this answers is not in the document]
Comment (dustin-miao) [reply solutions/advanced/cf-891C:14]: Just rephrase this to sound more professional (use more passive voice, parallelism, etc.).

File: solutions/advanced/cf-891C (hunk 10 above)
Comment (envyaims) [reply solutions/advanced/cf-891C:20]: wdym i didnt sort nothing

File: solutions/advanced/cf-891C (hunk 11 above)
[the remark this answers is not in the document]
Comment (envyaims) [reply solutions/advanced/cf-891C:14]: uhhh idk if its just me or it doenst make too much sense
@SansPapyrus683

File: solutions/advanced/cf-891C (hunk 10 above)
Comment (dustin-miao) [reply solutions/advanced/cf-891C:20]: Ok but you read in m + q numbers so that should definitely be in there?

Comment (envyaims) [reply solutions/advanced/cf-891C]: bruh fine

File: solutions/advanced/cf-891C (hunk 11 above)
[the remark this answers is not in the document]
Comment (SansPapyrus683) [reply solutions/advanced/cf-891C:14]: doesn't matter too much

**Comment by SansPapyrus683** [comment]
would like if you worked on some of the tasks in the github project
